### PR TITLE
[BEAM-593] Refactor Flink execution as preparation for async job control

### PR DESCRIPTION
**Pull request by aljoscha**
This is just the basic refactoring extracted from #2915. I think we should get this in rather quickly because several people seem to be working on Flink Runner stuff.

@peihe You might like to take a look, among other things I ported your recent additions on `FlinkPipelineExecutionEnvironment` (which doesn't exist anymore now) to the new `FlinkStreamingPipelineExecutor`.

### CODE REVIEW

**Comment by aljoscha** [comment]
CC: @JingsongLi

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPipelineExecutor.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  public PipelineResult executePipeline(
+      FlinkRunner runner, Pipeline pipeline, FlinkPipelineOptions options) throws Exception{
+
+    ExecutionEnvironment env = createBatchExecutionEnvironment(options);
+    FlinkBatchPipelineTranslator translator = new FlinkBatchPipelineTranslator(env, options);
+    translator.translate(pipeline);
+
+    JobExecutionResult result = env.execute(options.getJobName());
+
+    if (result instanceof DetachedEnvironment.DetachedJobExecutionResult) {
```
Comment (JingsongLi) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPipelineExecutor.java]: This is the same code as FlinkStreamingPipelineExecutor. Can we abstract it?

Comment (aljoscha) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPipelineExecutor.java]: You're right, I added a commit that refactors it into a util class.

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

**Comment by aljoscha** [comment]
I rebased on top of master and also made sure that the two-phase pipeline translation call is made in the right place, it's now called in `FlinkRunner` directly. @peihe this ports over your earlier change to another place, could you maybe check if I didn't break your changes?

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

**Comment by peihe** [comment]
Reviewed 3 of 6 files at r1, 3 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutor.java, line 20 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-KoqDdf4hQwUpbZ1Gz8q:-KoqDdf4hQwUpbZ1Gz8r:bx96l3a) ([raw file](https://github.com/apache/beam/blob/622b13f3ddb1ad6eb9c10897e3bc9bb1e5cfe8e5/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutor.java#L20)):*
> ```Java
> package org.apache.beam.runners.flink;
> ```

nit: remove extra line

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java, line 30 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-KotqzMwIBNMlG5MQy67:-KotqzMwIBNMlG5MQy68:b-f1km0b) ([raw file](https://github.com/apache/beam/blob/622b13f3ddb1ad6eb9c10897e3bc9bb1e5cfe8e5/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java#L30)):*
> ```Java
>  * {@link org.apache.beam.sdk.PipelineResult}.
>  */
> public class FlinkRunnerResultUtil {
> ```

drop public

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java, line 41 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-Kotr_OyINd30C34K-ue:-Kotr_OyINd30C34K-uf:b-1ougo) ([raw file](https://github.com/apache/beam/blob/622b13f3ddb1ad6eb9c10897e3bc9bb1e5cfe8e5/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java#L41)):*
> ```Java
>       if (accumulators != null && !accumulators.isEmpty()) {
>         log.info("Final accumulator values:");
>         for (Map.Entry<String, Object> entry : jobResult.getAllAccumulatorResults().entrySet()) {
> ```

-> accumulators.entrySet()

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineExecutor.java, line 140 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-KotvvUfZyXGXvdMNAbw:-KotvvUfZyXGXvdMNAbx:bzaafle) ([raw file](https://github.com/apache/beam/blob/622b13f3ddb1ad6eb9c10897e3bc9bb1e5cfe8e5/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineExecutor.java#L140)):*
> ```Java
>     JobExecutionResult jobResult = env.execute(options.getJobName());
> 
>     return FlinkRunnerResultUtil.wrapFlinkRunnerResult(LOG, jobResult);
> ```

is there going to be more code added to this method?

If not, maybe just inline two lines above in executePipeline()

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3295)*

**Comment by peihe** [comment]
Run Flink ValidatesRunner

**Comment by aljoscha** [comment]
Tests are currently failing, this will be fixed with #3550

**Comment by aljoscha** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutor.java, line 20 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-KoqDdf4hQwUpbZ1Gz8q:-Kr-tJJMV2dZTwsAmBFK:b-896fix) ([raw file](https://github.com/apache/beam/blob/39b294d3cf99a1ee14c4d50e8079597a745eba2c/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutor.java#L20)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

nit: remove extra line
</blockquote></details>

Done.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java, line 30 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-KotqzMwIBNMlG5MQy67:-Kr-tOPFLOG9T871IqLz:b-896fix) ([raw file](https://github.com/apache/beam/blob/39b294d3cf99a1ee14c4d50e8079597a745eba2c/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java#L30)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

drop public
</blockquote></details>

Done.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java, line 41 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-Kotr_OyINd30C34K-ue:-Kr-tWuJ5T229t54RXtH:b-896fix) ([raw file](https://github.com/apache/beam/blob/39b294d3cf99a1ee14c4d50e8079597a745eba2c/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResultUtil.java#L41)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

-> accumulators.entrySet()
</blockquote></details>

Done.

---

*[runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineExecutor.java, line 140 at r3](https://reviewable.io:443/reviews/apache/beam/3295#-KotvvUfZyXGXvdMNAbw:-Kr-tr9zGCAQIgTXO_lk:b-udjjv0) ([raw file](https://github.com/apache/beam/blob/39b294d3cf99a1ee14c4d50e8079597a745eba2c/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineExecutor.java#L140)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

is there going to be more code added to this method?

If not, maybe just inline two lines above in executePipeline()
</blockquote></details>

Yes, there will be added more code for the "async job submission" changes which were the original motivation for this refactoring.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3295)*

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

**Comment by aljoscha** [comment]
@peihe I addressed your comments and also rebased on master because this has become quite old.

**Comment by aljoscha** [comment]
The pre-commit hook failed because of a Gearpump failure.

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

**Comment by aljoscha** [comment]
retest this please

**Comment by aljoscha** [comment]
@peihe do you think this is good to merge now?

**Comment by jkff** [comment]
Is this PR still relevant?

**Comment by aljoscha** [comment]
I think it is, because it cleans up the code base and makes it easier to implement future stuff.

**Comment by iemejia** [comment]
This was marked as stale and would be closed if we don't accomplish progress, any way I can help you to get this finished/merged @aljoscha?

**Pull request closed without merging**